### PR TITLE
CMake: clang-cuda check

### PR DESCRIPTION
**Pull request by havogt**
Implements the second proposal in #1425

### CODE REVIEW

**Comment by havogt** [comment]
launch strgrid

**Comment by havogt** [comment]
Reminder: before merging, check that clang cuda is still detected on daint

**Comment by havogt** [comment]
launch strgrid

**Comment by havogt** [comment]
launch strgrid

**Comment by havogt** [comment]
launch strgrid

File: cmake/detect_features.cmake [hunk 1]
```diff
@@ -1,36 +1,97 @@
...
+            set(${gt_result} Clang-CUDA PARENT_SCOPE)
+            return()
+        endif()
+    endif()
+    set(${gt_result} NOTFOUND PARENT_SCOPE)
+endfunction()
+
+# try_nvcc_cuda()
+# Parameters:
+#    - gt_result: result variable is set to NVCC-CUDA  or NOTFOUND
```
Comment (havogt) [inline cmake/detect_features.cmake]:
```suggestion
#    - gt_result: result variable is set to NVCC-CUDA or NOTFOUND
```

File: cmake/try_compile_clang_cuda.cmake [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+#  try_compile_clang_cuda(CUDA_ARCH)
+#
+#  Parameters:
+#    - CUDA_ARCH: architecture used for testing compilation
+#                 (doesn't have to match the architecture of an available GPU, no code is executed)
+#
+# This function defines:
+#  GT_CLANG_CUDA_WORKS         CUDA test file was successfully compiled with clang
+#
+function(try_compile_clang_cuda CUDA_ARCH)
```
Comment (anstaf) [inline cmake/try_compile_clang_cuda.cmake]: Can you pass the name of the output variable as a parameter? (The same also for hip detection function.)

File: CMakeLists.txt [hunk 3]
```diff
@@ -25,16 +25,26 @@ if(NOT (CMAKE_CXX_COMPILER_ID STREQUAL AppleClang AND CMAKE_CXX_COMPILER_VERSION
...
+        enable_language(CUDA)
+    endif()
+
+    get_property(languages GLOBAL PROPERTY ENABLED_LANGUAGES)
+    if("CUDA" IN_LIST languages)
+        add_library(gridtools_nvcc INTERFACE)
+        # allow to call constexpr __host__ from constexpr __device__, e.g. call std::max in constexpr context
+        target_compile_options(gridtools_nvcc INTERFACE $<$<COMPILE_LANGUAGE:CUDA>:--expt-relaxed-constexpr>)
+    else()
+        message(STATUS "CUDA was not enabled. If you want CUDA support, please enable the language in your project")
```
Comment (anstaf) [inline CMakeLists.txt]: May be `message(WARNING)` here?

File: cmake/detect_features.cmake [hunk 4]
```diff
@@ -1,36 +1,97 @@
...
+                # Clang-CUDA doesn't work, try NVCC
+                try_nvcc_cuda(gt_result)
+                if(gt_result)
+                    set(${cuda_type} ${gt_result} PARENT_SCOPE)
+                    return()
+                endif()
 
-    set(${dst} NOTFOUND PARENT_SCOPE)
+                set(${cuda_type} NOTFOUND PARENT_SCOPE)
+                return()
```
Comment (anstaf) [inline cmake/detect_features.cmake]: `return()` is redundant here

File: CMakeLists.txt (hunk 3 above)
Comment (havogt) [inline CMakeLists.txt]:
```suggestion
        message(WARNING "CUDA was not enabled. If you want CUDA support, please enable the language in your project")
```

**Comment by havogt** [comment]
launch strgrid

**Pull request merged**